Minters and Burners

jgeary opened the pull request:
Add mapping of minters that `token.owner()` can update. Minters can also burn their own tokens.

Token is initialized such that mint and burn behavior is the same as previous Token implementation - the auction contract is the only minter and only burner at first.

jgeary commented:
I added burners too because identical logic. Both default to existing behavior - auction is only the minter and only burner at first. Token.owner() can then add and revoke minters and burners.

Looks like the `DeployTokenUpgrade.s.sol` is not quite finished so I should do that too as part of this PR, otherwise this is done.

tbtstl commented on src/token/Token.sol:
```diff
@@ -403,6 +417,46 @@ contract Token is IToken, UUPS, Ownable, ReentrancyGuard, ERC721Votes, TokenStor
+    /// @notice Update minters
+    /// @param _minters Array of addresses to add or remove as minters
+    /// @param _isMinter Array of booleans indicating whether the address is a minter
+    function updateMinters(address[] calldata _minters, bool[] calldata _isMinter) external onlyOwner {
```
👍

tbtstl commented on src/token/Token.sol:
```diff
@@ -403,6 +417,46 @@ contract Token is IToken, UUPS, Ownable, ReentrancyGuard, ERC721Votes, TokenStor
+            minter[_minters[i]] = _isMinter[i];
+
+            // Emit the event
+            emit MinterUpdated(_minters[i], _isMinter[i]);
```
might be a bit cleaner to make this `MintersUpdated(_minters, _isMinter)` and only emit the single event. Gas cost is trivial but I don't know what the advantage of many events would be.

iainnash replied:
+1

jgeary replied:
I think the individual event is preferable from a dumb indexer perspective but thats fine too. I'll change

jgeary replied:
With the individual minter update event, each event is guaranteed to contain new information (see a couple lines above - continue to next array element if minter already has the desired value). The entire array will likely contain redundant info due to people using excess caution and/or not entirely understanding the function.

tbtstl commented on src/token/Token.sol:
```diff
@@ -403,6 +417,46 @@ contract Token is IToken, UUPS, Ownable, ReentrancyGuard, ERC721Votes, TokenStor
+            burner[_burners[i]] = _isBurner[i];
+
+            // Emit the event
+            emit BurnerUpdated(_burners[i], _isBurner[i]);
```
same as above – `BurnersUpdated`

tbtstl commented on src/token/storage/TokenStorageV2.sol:
```diff
@@ -3,10 +3,10 @@ pragma solidity 0.8.16;
+/// @author Rohan Kulkarni and James Geary
 /// @notice The Token storage contract
-contract TokenStorageV1 is TokenTypesV1 {
+contract TokenStorageV2 is TokenTypesV1 {
```
See note above re: ephemeral storage

tbtstl commented on test/Token.t.sol:
```diff
@@ -363,23 +363,23 @@ contract TokenTest is NounsBuilderTest, TokenTypesV1 {
         assertEq(token.getPastVotes(founder, block.timestamp - 1), 2);
     }
 
-    function testRevert_OnlyAuctionCanMint() public {
```
Let's harden this test a bit, and turn it in to two. 

Firstly, let's make a separate test that the auction is a minter after deployment. That should cover the default case.

Secondly, let's make this a [fuzz test](https://book.getfoundry.sh/forge/fuzz-testing) that takes an address, updates the minter, and tries to mint.

tbtstl commented on test/Token.t.sol:
```diff
@@ -363,23 +363,23 @@ contract TokenTest is NounsBuilderTest, TokenTypesV1 {
         token.mint();
     }
 
-    function testRevert_OnlyAuctionCanBurn() public {
```
See above, we should make this into two test cases as well

tbtstl commented on test/Token.t.sol:
```diff
@@ -533,4 +533,126 @@ contract TokenTest is NounsBuilderTest, TokenTypesV1 {
         assertEq(mintedTokens[f3Wallet], f3Percentage);
     }
+
+    function test_updateMinters() public {
```
Also see above – this test is already decently hardened but let's just be extra extra careful here and make this a fuzz test as well. 

```solidity
function test_updateMinters(address m1, address m2, bool m1CanMint, bool m2CanMint) public {}
```

tbtstl commented on test/Token.t.sol:
```diff
@@ -533,4 +533,126 @@ contract TokenTest is NounsBuilderTest, TokenTypesV1 {
+        token.updateMinters(minters, isMinter);
+    }
+
+    function test_UpdateBurners() public {
```
Same as above, fuzz test would be ideal here.

tbtstl reviewed (commented):
This is a great first iteration – a couple of nits but looking very good

iainnash commented on src/token/Token.sol:
```diff
@@ -242,12 +261,7 @@ contract Token is IToken, UUPS, Ownable, ReentrancyGuard, ERC721Votes, TokenStor
-            revert ONLY_AUCTION();
-        }
-
+    function burn(uint256 _tokenId) external onlyBurner {
```
Do we have a design doc for this?

Not sure why we allow the auction to burn any token. It would make sense for me to allow only burns that are owned by the caller. 

Opening this burn type functionality owned by anyone to any address doesn't feel safe to me.

iainnash replied:
Another way to solve this is `AllowedMinter`s can burn only tokens held by the minter address.

With this code if I make an EOA the burner, I can eject anyone from the DAO with that EOA.

jgeary replied:
@iainnash any thoughts on what we should call this role? I think it's a good idea, but now "minter" seems insufficient if it can burn (in certain cases) as well. Maybe it's fine.

iainnash replied:
Shouldn’t burning just allow only the current holder to burn?
Do we enable standard burning across the board for existing tokens @tbtstl ?

jgeary replied:
this is outdated, see newest changes. its only minter that holds current token

iainnash commented on src/token/Token.sol:
```diff
@@ -403,6 +417,46 @@ contract Token is IToken, UUPS, Ownable, ReentrancyGuard, ERC721Votes, TokenStor
+    /// @notice Update minters
+    /// @param _minters Array of addresses to add or remove as minters
+    /// @param _isMinter Array of booleans indicating whether the address is a minter
+    function updateMinters(address[] calldata _minters, bool[] calldata _isMinter) external onlyOwner {
```
Thoughts on array of structs vs. two arrays here?

tbtstl replied:
+1, would solve the array length checking

jgeary replied:
sure

jgeary commented on src/token/Token.sol:
```diff
@@ -16,14 +16,35 @@ import { IToken } from "./IToken.sol";
 /// @author Rohan Kulkarni
 /// @notice A DAO's ERC-721 governance token
-contract Token is IToken, UUPS, Ownable, ReentrancyGuard, ERC721Votes, TokenStorageV1 {
+contract Token is IToken, UUPS, Ownable, ReentrancyGuard, ERC721Votes, TokenStorageV2 {
```
perhaps a better convention would be `contract TokenStorageV2 is TokenStorageV1` and we stick with `contract Token is ... TokenStorageV2` ?

tbtstl replied:
yea that's fair, let's do that

iainnash replied:
@jgeary One issue I see with this approach is:

1. `TokenStorageV3 is TokenStorageV2`
2. `TokenStorageV2 is TokenStorageV1`
3. `TokenStorageV1`

Feels more confusing to read in the future vs

`is TokenStorageV1, TokenStorageV2, TokenStorageV3`

in terms of reading a contract.

jgeary replied:
ok sounds good @iainnash

jgeary commented:
@tbtstl @iainnash check recent changes address all comments except tysons' re: tests

jgeary commented:
@iainnash @tbtstl this should be good but please give another look.

i went with iain's idea - there is no burner role, but a minter can burn a token that they own currently.

i added fuzzing for tests that tyson requested. in only fuzzed the addresses - but as for whether `bool authorized` is true or false, i split them into separate tests for adding and removing minters so i hardcoded those values.

iainnash commented on src/token/IToken.sol:
```diff
@@ -40,12 +45,24 @@ interface IToken is IUUPS, IERC721Votes, TokenTypesV1 {
+    error INVALID_MINTER_ARRAY_LENGTH();
+
+    /// @dev Reverts if burners and isBurner arrays are not the same length
+    error INVALID_BURNER_ARRAY_LENGTH();
```
Do we need these?

iainnash commented on src/token/Token.sol:
```diff
@@ -409,6 +424,21 @@ contract Token is IToken, VersionedContract, UUPS, Ownable, ReentrancyGuard, ERC
+            minter[_minters[i].minter] = _minters[i].authorized;
+        }
+
+        emit MintersUpdated(_minters);
```
Discussed with james and now that I've read the code and had context, it may make more sense to make this singular and then line 433 makes more sense.

iainnash replied:
Event proposal: `emit MinterUpdated(address minter, bool allowed)`

iainnash commented on script/DeployContracts.s.sol:
```diff
@@ -38,9 +44,14 @@ contract DeployContracts is Script {
 
         // Deploy root manager implementation + proxy
-        address managerImpl0 = address(new Manager(address(0), address(0), address(0), address(0), address(0)));
+        // address managerImpl0 = address(new Manager(address(0), address(0), address(0), address(0), address(0)));
```
Why is this commented out?

iainnash commented on script/DeployContracts.s.sol:
```diff
@@ -75,10 +86,10 @@ contract DeployContracts is Script {
 
-        console2.log("~~~~~~~~~~ MANAGER IMPL 0 ~~~~~~~~~~~");
-        console2.logAddress(managerImpl0);
+        // console2.log("~~~~~~~~~~ MANAGER IMPL 0 ~~~~~~~~~~~");
```
Should be safe to just delete

iainnash commented on src/token/Token.sol:
```diff
@@ -170,16 +198,17 @@ contract Token is IToken, VersionedContract, UUPS, Ownable, ReentrancyGuard, ERC
+        return _mintWithVesting(recipient);
+    }
 
+    function _mintWithVesting(address recipient) internal returns (uint256 tokenId) {
```
I like the new fn name

iainnash reviewed (changes requested):
Are we able to run auctions with upgraded implementations that do not have the auction address in the allowedMinters list?

Given that the auction is a common use case and we could save a memory write + an update upgrade step we may want to just always allow the auction to mint. It also breaks some assumptions we have around pausing and unpausing auctions, I believe we pause auctions if mint fails which would cause both a need to unpause the auction and fix mint (better than breaking entirely, but we should be cognizent of the upgrade path and failure cases here).

iainnash commented on src/token/Token.sol:
```diff
@@ -31,17 +31,17 @@ contract Token is IToken, VersionedContract, UUPS, Ownable, ReentrancyGuard, ERC
-    modifier onlyMinterWithToken(uint256 tokenId) {
-        if (!minter[msg.sender]) {
-            revert ONLY_MINTER();
+    modifier onlyAuctionOrMinterWithToken(uint256 tokenId) {
```
Could you just combine the two modifiers on burn? Or perhaps use the above modifier for burn then check the ownership within the burn function and not a modifier since it's only being used in one place.

iainnash commented on script/DeployContracts.s.sol:
```diff
@@ -17,6 +17,12 @@ import { ERC1967Proxy } from "../src/lib/proxy/ERC1967Proxy.sol";
+    // address constant TOKEN_120_GOERLI = 0xbD6F4BA8af93A1fBD0fd4993aD336dD9aB66F822;
+    // address constant TOKEN_120_MAINNET = 0xBE9Bd1e89E61C8Aae7d8bc24DD445ba120720D2d;
+
+    address constant MANAGER_MAINNET = 0xd310A3041dFcF14Def5ccBc508668974b5da7174;
```
Note to remove changes to deploy script.

iainnash reviewed (approved):
lgtm

noted by james is we may want to add these minting functions to the interface moving forward but this pr looks good.
